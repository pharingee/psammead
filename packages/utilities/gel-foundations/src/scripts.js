export arabic from './scripts/arabic';

export arabicPashto from './scripts/arabic-pashto';

export bengali from './scripts/bengali';

export burmese from './scripts/burmese';

export devanagariAndGurmukhi from './scripts/devanagari-and-gurmukhi';
export hindi from './scripts/devanagari-and-gurmukhi';
export nepali from './scripts/devanagari-and-gurmukhi';

export cyrillic from './scripts/latin-and-cyrillic';
export cyrillicAndLatin from './scripts/latin-and-cyrillic';
export latin from './scripts/latin-and-cyrillic';

export latinDiacritics from './scripts/latin-with-diacritics';

export chinese from './scripts/no-ascenders-or-descenders';
export ethiopic from './scripts/no-ascenders-or-descenders';
export korean from './scripts/no-ascenders-or-descenders';
export noAscendersOrDescenders from './scripts/no-ascenders-or-descenders';

export sinhalese from './scripts/sinhalese';

export tamil from './scripts/tamil';

export thai from './scripts/thai';
