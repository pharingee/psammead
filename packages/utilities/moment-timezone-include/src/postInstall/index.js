import writeNewTimezoneData from '../writeNewTimezoneData';

writeNewTimezoneData();
