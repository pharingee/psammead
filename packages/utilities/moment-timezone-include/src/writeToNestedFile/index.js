import fs from 'fs';
import mkdirp from 'mkdirp';
import path from 'path';

const writeToNestedFile = (filePath, contents) => {
  const fullFilePath = path.join(__dirname, filePath);
  const folder = path.dirname(fullFilePath);

  mkdirp.sync(folder);

  fs.writeFileSync(fullFilePath, contents);

  return fullFilePath;
};

export default writeToNestedFile;
