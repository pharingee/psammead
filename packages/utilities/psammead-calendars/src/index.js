export jalaali from './calendars/jalaali';
