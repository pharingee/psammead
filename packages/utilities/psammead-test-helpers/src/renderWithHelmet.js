import { render, waitForDomChange } from '@testing-library/react';

export default component => {
  /*
   * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
   * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
   * Pass in a component that uses helmet somewhere in the component tree.
   * The full html tree is returned which you can then use to snapshot helmet/head contents.
   */

  const { container } = render(component);
  const noop = () => {};
  const ARBITRARY_TIMEOUT = 100; // seems long enough for any dom mutations to occur
  const headElement = document.querySelector('head');

  headElement.innerHTML = ''; // clear out head mutations from previous tests

  return waitForDomChange({
    container: headElement,
    timeout: ARBITRARY_TIMEOUT,
  })
    .catch(noop) // handle a waitForDomChange timeout
    .then(mutationsList => {
      const headMutationDetected = mutationsList && mutationsList.length;

      if (headMutationDetected) {
        // helmet was probably used so we should get the full html

        const htmlElement = document.querySelector('html');

        const helmetElements = document.querySelectorAll(
          '[data-react-helmet="true"]',
        );

        const removeHelmetAttributes = el =>
          el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

        removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

        Array.from(helmetElements).forEach(removeHelmetAttributes);

        return { container: document.querySelector('html') };
      }

      return { container };
    });
};
