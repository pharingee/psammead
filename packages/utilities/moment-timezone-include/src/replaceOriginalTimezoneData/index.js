import rawTimezones from 'moment-timezone/data/packed/latest.json';
import writeToNestedFile from '../writeToNestedFile';

const replaceOriginalTimezoneData = () =>
  writeToNestedFile(
    `../../tz/${rawTimezones.version}-raw.json`,
    JSON.stringify({ ...rawTimezones, ...{ zones: [], links: [] } }),
  );

export default replaceOriginalTimezoneData;
