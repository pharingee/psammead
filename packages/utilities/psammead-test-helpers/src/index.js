import { render } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';
import renderWithHelmet from './renderWithHelmet';

export const shouldMatchSnapshot = (title, component) => {
  it(title, done => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    renderWithHelmet(component).then(({ container }) => {
      const hasOneChild = container.children.length === 1;
      /*
       * if the container has more than one child then it's a component that uses a
       * fragment at the top level so we should not select the first child because it
       * wouldn't snapshot the whole component
       */
      expect(
        hasOneChild ? removeWrappingDiv(container) : container,
      ).toMatchSnapshot();

      done();
    });
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
