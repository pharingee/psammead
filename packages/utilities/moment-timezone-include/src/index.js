import webpack from 'webpack';
import writeNewTimezoneData from './writeNewTimezoneData';
import replaceOriginalTimezoneData from './replaceOriginalTimezoneData';

function MomentTimezoneInclude(options = {}) {
  const startYear = options.startYear || -Infinity;
  const endYear = options.endYear || Infinity;

  writeNewTimezoneData(startYear, endYear);

  return new webpack.NormalModuleReplacementPlugin(
    /data[\\/]packed[\\/]latest\.json$/,
    resource => {
      resource.request = replaceOriginalTimezoneData(); // eslint-disable-line no-param-reassign
    },
  );
}

module.exports = MomentTimezoneInclude;
