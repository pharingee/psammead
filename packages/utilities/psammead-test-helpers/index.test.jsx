import React from 'react';
import Helmet from 'react-helmet';
import * as testHelpers from '.';
import * as testHelpersFromSrc from './src/index';
import renderWithHelmet from './src/renderWithHelmet';

const testHelpersExpectedExports = {
  shouldMatchSnapshot: 'function',
  isNull: 'function',
  testUtilityPackages: 'function',
  setWindowValue: 'function',
  resetWindowValue: 'function',
  suppressPropWarnings: 'function',
};

const expectedExports = { testHelpers: testHelpersExpectedExports };

const actualExports = { testHelpers };

const actualExportsFromSrc = { testHelpers: testHelpersFromSrc };

const serializeDomString = domString =>
  domString
    .split('\n')
    .map(line => line.trim())
    .join('')
    .trim();

const ensureErrorWhenMissingExport = testHelperMethod => {
  // missing export in the expected
  const actualWithAll = { utility: { foo: {}, bar: {} } };
  const expectedMissing = { utility: { bar: {} } };

  expect(() => {
    testHelperMethod(actualWithAll, expectedMissing, 'testing');
  }).toThrowError(
    "Missing value 'foo' in the expected export for 'testing/utility'.",
  );

  // missing export in the actual
  const actualMissing = { utility: { foo: {} } };
  const expectedWithAll = { utility: { bar: {}, foo: {} } };

  expect(() => {
    testHelperMethod(actualMissing, expectedWithAll, 'testing');
  }).toThrowError(
    "Missing value 'bar' in the actual export for 'testing/utility'.",
  );
};

const ensureErrorWhenMissingFileDefinition = testHelperMethod => {
  // missing export in the expected
  const actualWithAll = {
    utilityOne: { key: {} },
    utilityTwo: { key: {} },
  };
  const expectedMissing = {
    utilityOne: { key: {} },
  };

  expect(() => {
    testHelperMethod(actualWithAll, expectedMissing, 'testing');
  }).toThrowError(
    "Missing value 'utilityTwo' in the expected utilities for 'testing'.",
  );

  // missing export in the actual
  const actualMissing = {
    utilityOne: { key: {} },
  };
  const expectedWithAll = {
    utilityOne: { key: {} },
    utilityTwo: { key: {} },
  };

  expect(() => {
    testHelperMethod(actualMissing, expectedWithAll, 'testing');
  }).toThrowError(
    "Missing value 'utilityTwo' in the actual utilities for 'testing'.",
  );
};

describe('Psammead test helpers', () => {
  it('should test all the utility exports exist and are the correct type', () => {
    testHelpers.testUtilityPackages(
      actualExports,
      expectedExports,
      'psammead-test-helpers',
    );
  });

  it('should test all the utility exports exist and are the correct type when coming from src/', () => {
    testHelpers.testUtilityPackages(
      actualExportsFromSrc,
      expectedExports,
      'psammead-test-helpers',
    );
  });

  it('should error if expectedExport is missing an export compared to the actual export', () => {
    ensureErrorWhenMissingExport(testHelpers.testUtilityPackages);
  });

  it('should error if expectedExport is missing an export compared to the actual export when coming from /src', () => {
    ensureErrorWhenMissingExport(testHelpersFromSrc.testUtilityPackages);
  });

  it('should error if file definition is missing in the expectation', () => {
    ensureErrorWhenMissingFileDefinition(testHelpers.testUtilityPackages);
  });

  it('should error if file definition is missing in the expectation when coming from /src', () => {
    ensureErrorWhenMissingFileDefinition(
      testHelpersFromSrc.testUtilityPackages,
    );
  });

  const NoHelmet = () => (
    <main>
      <h1>Hello I am a test component</h1>
    </main>
  );

  it('should return correct HTML for components not using helmet', async () => {
    const actual = await renderWithHelmet(<NoHelmet />);
    const expected = serializeDomString(`
    <div>
      <main>
        <h1>Hello I am a test component</h1>
      </main>
    </div>
    `);

    expect(actual.container.outerHTML).toEqual(expected);
  });

  testHelpers.shouldMatchSnapshot(
    'should match the snapshot for the test component',
    <NoHelmet />,
  );

  const NoHelmetWithFragment = () => (
    <>
      <h1>Hello I am a test component</h1>
      <p>I am some test text.</p>
    </>
  );

  it('should return correct HTML for components not using helmet and wrapped with fragment', async () => {
    const actual = await renderWithHelmet(<NoHelmetWithFragment />);
    const expected = serializeDomString(`
    <div>
      <h1>Hello I am a test component</h1>
      <p>I am some test text.</p>
    </div>
    `);

    expect(actual.container.outerHTML).toEqual(expected);
  });

  testHelpers.shouldMatchSnapshot(
    'should match the snapshot for the test component',
    <NoHelmetWithFragment />,
  );

  const HelmetOnly = () => (
    <Helmet htmlAttributes={{ dir: 'rtl', lang: 'fa' }}>
      <title>Snapshot with helmet only</title>
      <meta name="test name" content="test content" />
      <script src="test.js" />
    </Helmet>
  );

  it('should return correct HTML for components using helmet only', async () => {
    const actual = await renderWithHelmet(<HelmetOnly />);
    const expected = serializeDomString(`
    <html dir="rtl" lang="fa">
      <head>
        <title>Snapshot with helmet only</title>
        <meta name="test name" content="test content"><script src="test.js"></script>
      </head>
      <body>
        <div></div>
      </body>
    </html>
    `);

    expect(actual.container.outerHTML).toEqual(expected);
  });

  testHelpers.shouldMatchSnapshot(
    'should match the snapshot for the test component with helmet only',
    <HelmetOnly />,
  );

  const HelmetWithContent = () => (
    <>
      <Helmet htmlAttributes={{ dir: 'rtl', lang: 'fa' }}>
        <title>Snapshot with helmet and and other content</title>
        <meta name="test name" content="test content" />
        <script src="test.js" />
      </Helmet>
      <main>
        <h1>Hello I am a test component with React Helmet</h1>
      </main>
    </>
  );

  it('should return correct HTML for components using helmet only with other content', async () => {
    const actual = await renderWithHelmet(<HelmetWithContent />);
    const expected = serializeDomString(`
    <html dir="rtl" lang="fa">
      <head>
        <title>Snapshot with helmet and and other content</title>
        <meta name="test name" content="test content">
        <script src="test.js"></script>
      </head>
      <body>
        <div>
          <main>
            <h1>Hello I am a test component with React Helmet</h1>
          </main>
        </div>
      </body>
    </html>
    `);

    expect(actual.container.outerHTML).toEqual(expected);
  });

  testHelpers.shouldMatchSnapshot(
    'should match the snapshot for the test component with helmet and other content',
    <HelmetWithContent />,
  );
});
