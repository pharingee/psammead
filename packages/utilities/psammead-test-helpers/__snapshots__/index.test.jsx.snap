// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Psammead test helpers should match the snapshot for the test component 1`] = `
<main>
  <h1>
    Hello I am a test component
  </h1>
</main>
`;

exports[`Psammead test helpers should match the snapshot for the test component 2`] = `
<div>
  <h1>
    Hello I am a test component
  </h1>
  <p>
    I am some test text.
  </p>
</div>
`;

exports[`Psammead test helpers should match the snapshot for the test component with helmet and other content 1`] = `
<html
  dir="rtl"
  lang="fa"
>
  <head>
    <title>
      Snapshot with helmet and and other content
    </title>
    <meta
      content="test content"
      name="test name"
    />
    <script
      src="test.js"
    />
  </head>
  <body>
    <div>
      <main>
        <h1>
          Hello I am a test component with React Helmet
        </h1>
      </main>
    </div>
  </body>
</html>
`;

exports[`Psammead test helpers should match the snapshot for the test component with helmet only 1`] = `
<html
  dir="rtl"
  lang="fa"
>
  <head>
    <title>
      Snapshot with helmet only
    </title>
    <meta
      content="test content"
      name="test name"
    />
    <script
      src="test.js"
    />
  </head>
  <body>
    <div />
  </body>
</html>
`;
